### ISSUE TITLE
error: failed to materialize conversion for result #0 of operation 'arith.constant' that remained live after conversion

### ISSUE THREAD

**Issue by silvasean** [iree-org/iree#9749]
### What happened?

```
/tmp/repro.mlir:2:3: error: failed to materialize conversion for result #0 of operation 'arith.constant' that remained live after conversion
  func.func @forward(%arg0: tensor<?x?x?xf32>) -> tensor<?x4x7xf32> {
  ^
/tmp/repro.mlir:2:3: note: see current operation: %17 = "arith.constant"() {value = 0 : index} : () -> index
/tmp/repro.mlir:30:11: note: see existing live user here: %141:3 = "flow.dispatch.default_workgroup_count"(%17, %77, %81) : (index, index, index) -> (index, index, index)
    %22 = tensor.insert_slice %14 into %21[0, 0, 0] [%17, %18, %19] [1, 1, 1] : tensor<?x?x?xf32> into tensor<?x?x?xf32>
```

Full error log: https://gist.github.com/silvasean/27f5a847ae58028020aec5d7904f3e73

### Steps to reproduce your issue

```
iree-compile --iree-hal-target-backends=dylib /tmp/repro.mlir
```

```
module attributes {torch.debug_module_name = "SliceStartEqEndModule"} {
  func.func @forward(%arg0: tensor<?x?x?xf32>) -> tensor<?x4x7xf32> {
    %c0 = arith.constant 0 : index
    %c2 = arith.constant 2 : index
    %cst = arith.constant 1.000000e+00 : f32
    %c5_i64 = arith.constant 5 : i64
    %c1 = arith.constant 1 : index
    %0 = tensor.dim %arg0, %c0 : tensor<?x?x?xf32>
    %1 = arith.index_cast %0 : index to i64
    %2 = arith.cmpi sgt, %c5_i64, %1 : i64
    %3 = arith.select %2, %1, %c5_i64 : i64
    %4 = arith.index_cast %3 : i64 to index
    %5 = tensor.dim %arg0, %c1 : tensor<?x?x?xf32>
    %6 = tensor.dim %arg0, %c2 : tensor<?x?x?xf32>
    %7 = tensor.extract_slice %arg0[%4, 0, 0] [0, %5, %6] [1, 1, 1] : tensor<?x?x?xf32> to tensor<0x?x?xf32>
    %8 = arith.cmpi sge, %5, %c0 : index
    %9 = arith.select %8, %5, %c0 : index
    %10 = tensor.extract_slice %7[0, 0, 0] [0, %9, %6] [1, 1, 1] : tensor<0x?x?xf32> to tensor<0x?x?xf32>
    %11 = arith.cmpi sge, %6, %c0 : index
    %12 = arith.select %11, %6, %c0 : index
    %13 = tensor.extract_slice %10[0, 0, 0] [0, %9, %12] [1, 1, 1] : tensor<0x?x?xf32> to tensor<0x?x?xf32>
    %14 = tensor.cast %13 : tensor<0x?x?xf32> to tensor<?x?x?xf32>
    %15 = linalg.init_tensor [1, 4, 7] : tensor<1x4x7xf32>
    %16 = linalg.fill ins(%cst : f32) outs(%15 : tensor<1x4x7xf32>) -> tensor<1x4x7xf32>
    %17 = tensor.dim %14, %c0 : tensor<?x?x?xf32>
    %18 = tensor.dim %14, %c1 : tensor<?x?x?xf32>
    %19 = tensor.dim %14, %c2 : tensor<?x?x?xf32>
    %20 = arith.addi %17, %c1 : index
    %21 = linalg.init_tensor [%20, %18, %19] : tensor<?x?x?xf32>
    %22 = tensor.insert_slice %14 into %21[0, 0, 0] [%17, %18, %19] [1, 1, 1] : tensor<?x?x?xf32> into tensor<?x?x?xf32>
    %23 = tensor.insert_slice %16 into %22[%17, 0, 0] [1, 4, 7] [1, 1, 1] : tensor<1x4x7xf32> into tensor<?x?x?xf32>
    %24 = tensor.cast %23 : tensor<?x?x?xf32> to tensor<?x4x7xf32>
    return %24 : tensor<?x4x7xf32>
  }
}
```

### What component(s) does this issue relate to?

Compiler

### Version information

ca51bc61ab5a921b4150faf64422e4f92b2325cb

**Comment by MaheshRavishankar** [comment]
@silvasean again this is having a `tensor<0x?x?xf32>` .... Is that intended?

**Comment by silvasean** [comment]
Yes, the test specifically checks a slice with the start and end being equal.

```
class SliceStartEqEndModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    @export
    @annotate_args([
        None,
        ([-1, -1, -1], torch.float32, True),
    ])
    def forward(self, x):
        # TODO: remove hacky cat tensor once refbackend supports 0 size dim
        result = x[5:5, :, :]
        cat_tensor = torch.ones((1,4,7), dtype=torch.float32)
        return torch.cat((result, cat_tensor), dim=0)


@register_test_case(module_factory=lambda: SliceStartEqEndModule())
def SliceStartEqEndModule_basic(module, tu: TestUtils):
    module.forward(tu.rand(6,4,7))
```

**Comment by silvasean** [comment]
The comment there about 0 size dimensions is purely related to the calling convention -- the refbackend does the calculation just fine. The refbackend just lowers to loops and bufferizes.

**Comment by MaheshRavishankar** [comment]
I have a fix for this, but there is a need to fix this earlier. Such no-ops need to be pruned way ahead of time. 

@jpienaar we probably need a pass in Flow that can DCE such operations with zero-dim tensors. I cant think this on right now, but this should be an easy one. @okkwon does this sound like something you want to pick up?

**Comment by jpienaar** [comment]
The previous proposal was a pass to convert to these to linalg.init ops I believe, or can we just DCE?

**Comment by MaheshRavishankar** [comment]
> The previous proposal was a pass to convert to these to linalg.init ops I believe, or can we just DCE?

We should ideally just DCE them. I think `linalg.init_tensor` was a way to break dependency chains.

**Comment by benvanik** [comment]
Yeah DCE is always better, though we'll still need to support getting the empty tensors to `flow.tensor.empty` for when they are are live (returned from the program/stored into globals). `linalg.init_tensor` should already end up there thanks to Mahesh's InitializeEmptyTensorsPass so we just need to make those.
